contracts/delegation/dao-vote-delegation/src/helpers.rs

```diff
@@ -0,0 +1,391 @@
+    }
+
+    Ok(())
+}
+
+/// Validates a delegation and returns the delegator's voting power if valid.
+pub fn validate_delegation(
+    deps: Deps,
+    env: &Env,
+    delegator: &Addr,
+    delegate: &Addr,
+    new_percent: Decimal,
+) -> Result<Uint128, ContractError> {
+    if new_percent <= Decimal::zero() || new_percent > Decimal::one() {
+        return Err(ContractError::InvalidVotingPowerPercent {});
+    }
+
+    // delegates cannot delegate to others
+    if is_delegate_registered(deps, delegator, None)? {
+        return Err(ContractError::DelegatesCannotDelegate {});
+    }
+
+    // ensure delegate is registered
+    if !is_delegate_registered(deps, delegate, None)? {
+        return Err(ContractError::DelegateNotRegistered {});
+    }
+
+    // ensure delegator has voting power in the DAO
+    let vp = get_voting_power(
+        deps,
+        delegator,
+        // use next block height since voting power takes effect at the start of
+        // the next block. if the delegator changed their voting power in the
+        // current block, we need to use the new value.
+        env.block.height + 1,
+    )?;
+    if vp.is_zero() {
+        return Err(ContractError::NoVotingPower {});
+    }
+
+    Ok(vp)
+}
+
+/// Handle redelegation by updating the existing delegation.
+#[allow(clippy::too_many_arguments)]
+pub fn handle_redelegation(
+    deps: DepsMut,
+    env: &Env,
+    delegator: &Addr,
+    delegate: &Addr,
+    new_percent: Decimal,
+    config: &Config,
+    current_percent_delegated: Decimal,
+    vp: Uint128,
+    existing_delegation_entry: SnapshotVectorMapItemRef,
+) -> DelegationHandlerResult {
+    let (existing_delegation_id, existing_delegation_expiration) = existing_delegation_entry;
+    let expired = existing_delegation_expiration.is_some_and(|exp| exp <= env.block.height);
+
+    let existing_delegation_percent = DELEGATIONS
+        .load_item(deps.storage, delegator, existing_delegation_id)?
+        .percent;
+
+    // if delegation is not expired and percent is the same, just extend the
+    // expiration based on the current config.
+    if !expired && existing_delegation_percent == new_percent {
+        // if both expirations are none, do nothing.
+        if existing_delegation_expiration.is_none() && config.delegation_validity_blocks.is_none() {
+            return Ok((
+                current_percent_delegated,
+                (existing_delegation_id, existing_delegation_expiration),
+                false,
+            ));
+        }
+
+        let (_, new_expiration) = DELEGATIONS.update_expiration(
+            deps.storage,
+            delegator,
+            existing_delegation_id,
+            env.block.height,
+            config.delegation_validity_blocks,
+        )?;
+
+        update_delegated_vp_expiration(
+            deps.storage,
+            env,
+            delegate,
+            vp,
+            existing_delegation_expiration,
+            new_expiration,
+        )?;
```

Comment:
passing the `update_expiration` return value as the input for further calls is great!